src/core/string_map.cc

```diff
@@ -276,6 +276,24 @@ uint32_t StringMap::ObjExpireTime(const void* obj) const {
   return UINT32_MAX;
 }
 
+void StringMap::ObjUpdateExpireTime(const void* obj, uint32_t ttl_sec) {
+  sds str = (sds)obj;
+  char* valptr = str + sdslen(str) + 1;
+  uint32_t at = time_now() + ttl_sec;
+  absl::little_endian::Store32(valptr + 8, at);
+}
+
+void StringMap::ObjReplace(const void* obj, uint32_t ttl_sec) {
```

Comment:
I understand what you are trying to do but I think it can be simplified.

1. What you really want is `AddExpiry` semantics on the EXISTING object that DOES NOT have expiry. 
2. Which means you already have the iterator to the object, or even stronger the pointer to the hashtable entry/Link that contain that existing object pointer. 
3. So you do not need to do the whole AddOrReplaceObj flow again - you already have the position (DensePtr). All you need to do is to "clone" the existing object with TTL attached to it.
4. Once you do that, you can swap the pointers inside DensePtr and update DensePtr tag bits to add the meta information about TTL.

Reply:
Finally you will need to free the previous object pointer.

Reply:
I recently added `ObjectClone(const void* obj, bool has_ttl) ` that clones the object while keeping its ttl property as is.

You can extend its functionality to support `bool add_ttl` as well, so that `ObjectClone(existing_obj. has_ttl=false, add_ttl=true)` will clone the object that did not have ttl to the object that has (0) ttl and then you can call `ObjUpdateExpireTime` to update it to correct value

Reply:
Two things:

1. I do not personally like adding (yet) another boolean in `ObjectClone`. One bool is fine (usually), but at the moment you start adding different bools to an interface things become less and less clear.
2. `clone + swap` is almost identical to `AddOrReplace` when the second behaves as a `Replace`.

Also, notice that the logic of `ObjReplace` is again *almost* identical to the implementation of `AddOrUpdate()`. Also, notice that `AddOrUpdate` exists in: `StringMap` and `ScoreMap` (with the former having a slightly different return type). And finally `AddOrUpdate` is missing from `StringSet`. So why don't we make `AddOrUpdate` a `virtual` function ? The only thing we need to change is a) the signature of `StringMap::AddOrUpdate` to return a pair (just like `ScoredMap` -- and it's not called in many places so the change is quick) and implement this in `StringSet`. Now, we can get rid `ObjReplace` alltogether.

That way `SetExpiryTime` can become a member of `DenseSet` (and not a member of `DenseSet::Iterator`) + keep it in the base class to avoid duplicating in the derived classes (bonus points is also generic as you suggested).

@romange WDYT ? 

P.s. not a strong opinion, just discussing my thoughts

Reply:
Thanks for feedback from the both of you :)

My 2 cents: It seems like you're seeing the problem from different angles perhaps. One leaning towards code cleanliness, and the other performance.

I think `AddOrUpdate` isn't implemented in `StringSet` yet, because `Update` is less of a thing in a `Set` than it would be in `StringMap` or `ScoreMap`. And by that metric, we would probably want it to be a private method thats somehow shaded to be public for `StringMap` and `ScoreMap` if we went that route?

I do appreciate the desire to make the code-path hit only as few lines as possible, since we're only supposed to interact with already existing objects in this case. Too many flags in `ObjectClone` makes it a bit nasty, but skimming the code, the complexity doesn't increase that much internally at least.

Reply:
This is my best suggestion so far, which tries to keep the complexity of additional branching from `add_ttl` as low as possible: 
```cpp
void* StringSet::ObjectClone(const void* obj, bool has_ttl, bool add_ttl) const {
  sds src = (sds)obj;
  string_view sv{src, sdslen(src)};
  if(add_ttl){
  return (void*)MakeSetSds(sv, 0);
  }
  uint32_t at = has_ttl ? ObjExpireTime(obj) : UINT32_MAX;
  return (void*)MakeSetSds(sv, at);
}

sds StringSet::MakeSetSds(string_view src, uint32_t ttl_sec) const {
  DCHECK_GT(ttl_sec, 0u);  // ttl_sec == 0 would mean find and delete immediately
  if (ttl_sec != UINT32_MAX) {
    uint32_t at = time_now() + ttl_sec;
    DCHECK_LT(time_now(), at);

    sds newsds = AllocImmutableWithTtl(src.size(), at);
    if (!src.empty())
      memcpy(newsds, src.data(), src.size());
    return newsds;
  }

  return sdsnewlen(src.data(), src.size());
}
```

Reply:
Let's wait for Roman to decide what he likes the most. I do not have a strong preference :)